cmake/std/pr_python_test_driver.sh

```diff
@@ -0,0 +1,100 @@
+#!/usr/bin/env bash
+
+#
+# Prepare Environment
+#
+
+# The existence of PYTHONHOME can cause some odd problems
+# if we're seeing errors that look like `No module named 'encodings'`
+# then a usual fix is to try and unset PYTHONHOME.
+# I think this is due to a conflict in modules loaded by SEMS and then
+# our use of a specific Python that's different.
+# - See: https://stackoverflow.com/q/38132755/2059999 for more information
+unset PYTHONHOME
+
+
+#
+# Get pip
+# - @param1 python_exe - the python executable to install PIP for
+function get_pip() {
+    local python_exe=${1:?}
+
+    echo -e "--- Python: ${python_exe:?}"
+
+    # fetch get-pip.py
+    echo -e "--- curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py"
+    curl https://bootstrap.pypa.io/get-pip.py -o get-pip.py
+
+    get_pip_args=(
+        --user
+        --proxy="http://wwwproxy.sandia.gov:80"
+        --no-setuptools
+        --no-wheel
+    )
+    echo -e ""
+    echo -e "--- ${python_exe:?} ./get-pip.py ${get_pip_args[@]}"
+    ${python_exe:?} ./get-pip.py ${get_pip_args[@]}
+}
+
+
+#
+# Install Python pacakges using pip
+#
+# - @param1 pip_exe - the pip binary to use, i.e., pip3.
+#
+function get_python_packages() {
+    local pip_exe=${1:?}
+
+    echo -e "--- Pip   : ${pip_exe:?}"
+
+    pip_args=(
+        --use-feature=2020-resolver
+        configparser
+        mock
+        pytest
+        pytest-cov
+    )
+    echo -e "--- ${pip_exe:?} install --user ${pip_args[@]}"
+    ${pip_exe:?} install --user ${pip_args[@]}
+}
+
+
+
+echo "+================================================================================+"
+echo "|"
+echo "|                   P Y T H O N   T E S T   D R I V E R"
+echo "|"
+echo "+================================================================================+"
+echo ""
+echo "+--------------------------------------+"
+echo "|   P A C K A G E   S E T U P"
+echo "+--------------------------------------+"
+: "${1:?Param 1 ERROR: Python Executable not set or empty}"
+: "${2:?Param 2 ERROR: Pip Executable not set or empty}"
```

Comment:
I've never seen this pattern before.  What's happening here?

Reply:
If the parameter ${1} or ${2} don't exist, then an error is triggered with the message.

for example:
```bash
#!/usr/bin/env bash

: "${1:?Param 1 is missing or empty}"

echo "Param 1 = ${1}"
```

without the parameter:
```
$ ~/param_check.sh
/Users/wcmclen/param_check.sh: line 3: 1: Param 1 is missing or empty
```

with the parameter
```
$ ~/param_check.sh FOO
Param 1 = FOO
```

Think of using this construct as a way to validate expected parameters at the start of a script... kind of a 'design by contract' construct.

Reply:
Oh, nice.  Yeah I'd never seen this method of validating parameters in bash.

Reply:
Can I get that shaking fist emoji for bash please?